File: wisefy/src/androidTest/java/com/isupatches/wisefy/BaseAndroidJUnit4TestClass.java
package com.isupatches.wisefy;


import android.net.ConnectivityManager;
import android.net.wifi.WifiManager;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import org.junit.After;
import org.junit.Before;
import org.junit.runner.RunWith;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;


@RunWith(AndroidJUnit4.class)
public abstract class BaseAndroidJUnit4TestClass {

    ConnectivityManager mMockConnectivityManager;

    protected WifiManager mMockWiFiManager;

    protected WiseFy mWiseFy;

    protected static final Integer VERIFICATION_SUCCESS_TIMEOUT = 5000;

    static final Integer VERIFICATION_FAILURE_TIMEOUT = 3000;

    @Before
    public void setUp() {
        mMockWiFiManager = mock(WifiManager.class);
        mMockConnectivityManager = mock(ConnectivityManager.class);

        mMockWiFiManager = mock(WifiManager.class);
        mMockConnectivityManager = mock(ConnectivityManager.class);

        mWiseFy = new WiseFy.brains(InstrumentationRegistry.getContext()).logging(true).getSmarts();
        setManagers();
    }

    @After
    public void tearDown() {
        mWiseFy.dump();
    }

    /**
     * HELPERS
     */

    public void missingPrerequisite() {
        WiseFyPrerequisites mockPrereqs = mock(WiseFyPrerequisites.class);
        mWiseFy.mWiseFyPrerequisites = mockPrereqs;
        when(mockPrereqs.hasPrerequisites()).thenReturn(false);
    }

    void networkAlreadyInConfigurationList() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.isNetworkASavedConfiguration(anyString())).thenReturn(true);
    }

    private void setManagers() {
        WiseFyPrerequisites mockPrereqs = mock(WiseFyPrerequisites.class);
        mWiseFy.mWiseFyPrerequisites = mockPrereqs;
        mWiseFy.mWiseFySearch.mWiseFyPrerequisites = mockPrereqs;
        when(mockPrereqs.hasPrerequisites()).thenReturn(true);
        when(mockPrereqs.getWifiManager()).thenReturn(mMockWiFiManager);
        when(mockPrereqs.getConnectivityManager()).thenReturn(mMockConnectivityManager);
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/BaseAndroidJUnit4TestClass.kt
package com.isupatches.wisefy

import android.net.ConnectivityManager
import android.net.wifi.WifiConfiguration
import android.net.wifi.WifiManager
import androidx.test.platform.app.InstrumentationRegistry

import com.isupatches.wisefy.callbacks.AddNetworkCallbacks
import com.isupatches.wisefy.connection.WiseFyConnection
import com.isupatches.wisefy.search.WiseFySearch

import org.junit.After

import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito
import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * A common test class to extend.
 *
 * @author Patches
 */
internal open class BaseAndroidJUnit4TestClass {

    protected val wisefy: WiseFy
    protected val mockWiseFySearch: WiseFySearch

    protected val mockConnectivityManager: ConnectivityManager
    protected val mockWifiManager: WifiManager

    protected val mockNetworkUtil: MockNetworkUtil
    protected val mockWiseFyConnectionUtil: MockWiseFyConnectionUtil
    protected val mockWiseFyPrechecksUtil: MockWiseFyPrechecksUtil
    protected val mockWiseFySearchUtil: MockWiseFySearchUtil

    protected val nullCallbackUtil: NullCallbackUtil

    protected val verificationUtil: VerificationUtil

    /**
     * Constructor.
     */
    init {
        val mockWiseFyConnection = mock(WiseFyConnection::class.java)
        val mockWiseFyPrechecks = mock(WiseFyPrechecks::class.java)
        mockWiseFySearch = mock(WiseFySearch::class.java)

        mockWifiManager = mock(WifiManager::class.java)
        mockConnectivityManager = mock(ConnectivityManager::class.java)

        wisefy = WiseFy.Brains(InstrumentationRegistry.getInstrumentation().targetContext)
            .customConnectivityManager(mockConnectivityManager)
            .customWifiManager(mockWifiManager)
            .customWiseFyConnection(mockWiseFyConnection)
            .customWiseFyPrechecks(mockWiseFyPrechecks)
            .customWiseFySearch(mockWiseFySearch)
            .logging(true)
            .getSmarts()

        mockNetworkUtil = MockNetworkUtil(mockConnectivityManager, mockWifiManager)

        mockWiseFyConnectionUtil = MockWiseFyConnectionUtil(mockWiseFyConnection)
        mockWiseFyPrechecksUtil = MockWiseFyPrechecksUtil(mockWiseFyPrechecks)
        mockWiseFySearchUtil = MockWiseFySearchUtil(mockWiseFySearch)

        nullCallbackUtil = NullCallbackUtil(wisefy)

        verificationUtil = VerificationUtil(mockConnectivityManager, mockWifiManager)
    }

    @After fun tearDown() {
        wisefy.dump()
    }

    /*
     * HELPERS
     */

    protected fun <T> any(type: Class<T>): T = Mockito.any<T>(type)

    protected fun verifyNetworkAdded(mockCallbacks: AddNetworkCallbacks) {
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkAdded(anyInt(), any(WifiConfiguration::class.java))
    }
}
